runtime/lua/vim/lsp.lua

```diff
@@ -1704,6 +1705,14 @@ function lsp.start_client(config)
       end
     end)
 
+    -- Enable pull diagnostics for the buffer if the server supports it.
+    if vim.tbl_get(client.server_capabilities, 'diagnosticProvider') then
```

Comment:
This should probably use `supports_method` instead, to cover the dynamic capability.
And maybe run this within `lsp._set_defaults`, because that's already called when dynamic capabilities are updated.